File: src/leetcode/0151-Reverse-Words-in-a-String/solution.cpp
// Copyright (c) 2019 shaqsnake. All rights reserved.
/**
 * @Author: shaqsnake
 * @Email: shaqsnake@gmail.com
 * @Date: 2019-08-12 14:22:31
 * @LastEditTime: 2019-08-12 15:07:15
 * @Description: 151. Reverse Words in a String
 */

class Solution {
public:
    string reverseWords(string s) {
        int i = 0, k = 0;
        while (i < s.size()) {
            while (s[i] == ' ') i++;
            if (i == s.size()) break;

            int j = i;
            while (j < s.size() && s[j] != ' ') j++;
            
            reverse(s.begin() + i, s.begin() + j);

            if (k) s[k++] = ' ';
            while (i < j) s[k++] = s[i++];
        }

        s.erase(s.begin() + k, s.end());
        reverse(s.begin(), s.end());

        return s;
    }
};

File: src/leetcode/0151-Reverse-Words-in-a-String/solution.py
'''
@Author: shaqsnake
@Email: shaqsnake@gmail.com
@Date: 2019-05-09 13:47:55
@LastEditTime: 2019-08-12 15:00:13
@Description: 151. Reverse Words in a String
'''

class Solution(object):
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        return ' '.join(s.split()[::-1])

if __name__ == '__main__':
    s = Solution().reverseWords("the sky is blue")
    print(s)